File: java/src/main/java/topics/graph/BFS.java
package topics.graph;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * 너비 우선 탐색(BFS) 템플릿.
 *
 * 목표:
 * - 시작 정점에서 도달 가능한 정점을 방문 순서대로 반환한다.
 * - 시간복잡도 O(V + E)로 동작한다.
 *
 * 계약:
 * - graph: 각 정점의 인접 정점 목록을 담은 인접 리스트
 * - start: 시작 정점 인덱스
 * - 반환: 방문 순서를 담은 int 배열
 * - 입력이 유효하지 않으면 IllegalArgumentException을 던진다.
 */
public class BFS {
    public int[] bfs(int[][] graph, int start) {
        if(graph == null || start<0|| start >= graph.length )
            throw new IllegalArgumentException("지원하지 않는 입력입니다.");

        Queue<Integer> vertices=  new LinkedList<>(List.of(start));
        boolean[] visitedVertices = new boolean[graph.length];
        ArrayList<Integer> ret = new ArrayList<>();

        while(!vertices.isEmpty()){
            int vertex = vertices.poll();
            if(visitedVertices[vertex]) continue;
            ret.add(vertex);
            visitedVertices[vertex] = true;

            for(int v : graph[vertex]){
              if(v<0|| v >= graph.length) continue;
              vertices.add(v);
            }
        }

        int[] result = new int[ret.size()];
        
        for(int i = 0 ; i < ret.size(); i++)
            result[i] = ret.get(i);

        return result;
    }
}

File: javascript/src/graph/bfs.js
/**
 * 너비 우선 탐색(BFS) 템플릿.
 *
 * 목표:
 * - 시작 정점에서 도달 가능한 정점들을 레벨 순서로 방문
 * - 시간복잡도 O(V + E)
 *
 * 입력 계약:
 * - graph: 각 정점의 인접 정점 목록을 담은 인접 리스트
 * - start: 시작 정점 인덱스
 * - 잘못된 입력이면 예외를 던짐
 *
 * 구현 가이드:
 * 1. 입력과 시작점 범위를 검증한다.
 * 2. 큐와 방문 배열을 초기화한다.
 * 3. 시작 정점에서 인접 정점을 레벨 순서로 확장해 방문 순서를 기록한다.
 * 4. 방문 순서를 담은 배열을 반환한다.
 */
function bfs(graph, start) {
  if (!Array.isArray(graph) || start < 0 || start >= graph.length)
    throw new Error("지원하지 않는 입력입니다.");

  const visitedVertices = new Array(graph.length).fill(false);
  const [ret, queue] = [[],[start]];
  let queueHead = 0;
  ret.push(start);
  visitedVertices[start] = true;

  while (queueHead < queue.length){
    const vertex = queue[queueHead++];

    for (const v of graph[vertex]){
      if (v < 0 || v >= graph.length || visitedVertices[v]) continue;
      queue.push(v);
      ret.push(v);
      visitedVertices[v] = true;
    }
  }

  return ret;
}

module.exports = { bfs };
